pkg/encryption/elgamal/encryption.go

```diff
@@ -0,0 +1,189 @@
+package elgamal
+
+import (
+	crand "crypto/rand"
+	"fmt"
+	"github.com/bwesterb/go-ristretto"
+	"github.com/coinbase/kryptology/pkg/core/curves"
+	"math/big"
+)
+
+type TwistedElGamal struct {
+	curve   *curves.Curve
+	mapping map[string]uint64
+
+	// Notes the sizes decrypted so far. This helps us know if the values we need are already in the map.
+	maxMapping map[MaxBits]bool
+}
+
+func NewTwistedElgamalWithED25519Curve() *TwistedElGamal {
+	var s ristretto.Point
+	s.SetZero()
+	mapping := make(map[string]uint64)
+	maxMapping := make(map[MaxBits]bool)
+	mapping[s.String()] = 0
+
+	return &TwistedElGamal{
+		curve:      curves.ED25519(),
+		maxMapping: maxMapping,
+		mapping:    mapping,
+	}
+}
+
+func NewTwistedElgamal(curve *curves.Curve) *TwistedElGamal {
+	var s ristretto.Point
+	s.SetZero()
+	mapping := make(map[string]uint64)
+	maxMapping := make(map[MaxBits]bool)
+	mapping[s.String()] = 0
+
+	return &TwistedElGamal{
+		curve:      curve,
+		maxMapping: maxMapping,
+		mapping:    mapping,
+	}
+}
+
+// Encrypt encrypts a message using the public key pk.
+func (teg TwistedElGamal) Encrypt(pk curves.Point, message uint64) (*Ciphertext, curves.Scalar, error) {
+	// Generate a random scalar r
+	randomFactor := teg.curve.Scalar.Random(crand.Reader)
+
+	return teg.EncryptWithRand(pk, message, randomFactor)
+}
+
+// EncryptWithRand encrypts a message using the public key pk and a given random factor.
+func (teg TwistedElGamal) EncryptWithRand(pk curves.Point, message uint64, randomFactor curves.Scalar) (*Ciphertext, curves.Scalar, error) {
```

Comment:
Actually would recommend not exposing this function - it was added to test certain properties about the proofs, but this should never be used to encrypt stuff since encryption should be done with a random value to be secure.

Reply:
Made it private for now.